lib/foreman_tasks_core/runner/action.rb

```diff
@@ -52,15 +54,34 @@ def finish_run(update)
         output[:exit_status] = update.exit_status
       end
 
+      def process_external_event(event)
+        runner_dispatcher.external_event(output[:runner_id], event)
+        suspend
+      end
+
       def process_update(update)
-        output[:result].concat(update.continuous_output.raw_outputs)
+        merge_outputs(update.continuous_output.raw_outputs, update.incremental)
         if update.exit_status
           finish_run(update)
         else
           suspend
         end
       end
 
+      def merge_outputs(outputs, incremental)
+        # No need to do anything if there are no inputs or if the "new" outputs are older than the already saved ones
+        return if outputs.empty? || (!output[:result].empty? && outputs.first['timestamp'] < output[:result].last['timestamp'])
```

Comment:
the `outputs` is quite confusing with the `output`, something like `new_chunks` and extracting `output[:result]` to `known_chunks`, (or something better than `chunks`:) would help reading this code